src/server.c

```diff
@@ -677,6 +677,42 @@ int clientsCronResizeQueryBuffer(client *c) {
     return 0;
 }
 
+/* The client output buffer can be adjusted to better fit the memory requierments.
+ *
+ * the logic is:
+ * in case the last observed peak size of the buffer equals the buffer size - we double the size
+ * in case the last observed peak size of the buffer equals less than half the buffer size - we shrink by half.
+ * in any case we will reset the peak value each cron run.
+ * The function always returns 0 as it never terminates the client. */
+int clientsCronResizeOutputBuffer(client *c) {
+
+    void *new_buf = NULL;
+
+    if(c->bufpos != 0)  return 0;
+
+    if (c->buf_usable_size > PROTO_REPLY_MIN_BYTES &&
+        c->buf_peak < c->buf_usable_size/2 ) {
+        new_buf = ztryrealloc(c->buf,c->buf_usable_size/2);
+        server.reply_buffer_shrinks++;
+
+    }
+    else if(c->buf_usable_size < PROTO_REPLY_CHUNK_BYTES &&
+            c->buf_peak == c->buf_usable_size) {
```

Comment:
This looks a little odd, buf_peak will never be exactly equal to buf_usable_size, and buf does not seem to be shrunk normally.

Reply:
@sundb we usually try to write as much as possible to the client reply buffer and if filled completely (c->bufpos == c->buf_usable_size) we continue to reply list allocations.
Regarding the shrinking - I am not sure I follow the problem...
maybe you can elaborate?

Reply:
i think Ran is right.. these days we try to fill it to the last byte (it wasn't like that in the past).
however, i'm afraid that the *2 and /2 thing can somehow cause a resonance effect, especially considering that we don't get the size we asked for (we can get up to 30% more due to internal frag).
maybe we should go with *2 and /4 ?

Reply:
@oranagra 
we will toggle these sizes [1024,2048,4096,8Kib,16Kib]
I agree that in case these values will change in the future the code would be potentially 25%-30% extra size (like the jump from 8Kib size class to 10Kib)  but we will still make use of the extra size in our attempt to shrink and expend the buffer.
I just feel 4 times shrinkage might be too aggressive. 
Also I thought a lot about the 1Kib min limit which is basically 2 times larger than my plan. since our analysis of customer data sizes shows most cases value sizes are up to 512 bytes.

Reply:
ok.. i don't think the current code can actually cause a resonance effect. it's just a paranoia i have of two thresholds being too close to each other. but i suppose in this case it's ok